src/core/kdtree.hpp

```diff
@@ -0,0 +1,457 @@
+
+		if (next == INVALID_NODE) {
+			/* New leaf */
+			size_t newidx = this->AddNode(element);
+			/* Vector may have been reallocated at this point, n and next are invalid */
+			auto &nn = this->nodes[node_idx];
+			if (ec < nc) nn.left = newidx; else nn.right = newidx;
+		} else {
+			this->InsertRecursive(element, next, level + 1);
+		}
+	}
+
+	/**
+	 * Free all children of the given node
+	 * @return Collection of elements that were removed from tree.
+	 */
+	std::vector<T> FreeSubtree(size_t node_idx)
+	{
+		std::vector<T> subtree_elements;
+		auto &n = this->nodes[node_idx];
+
+		/* We'll be appending items to the free_list, get index of our first item */
+		size_t first_free = this->free_list.size();
+		/* Prepare the descent with our children */
+		if (n.left != INVALID_NODE) this->free_list.push_back(n.left);
+		if (n.right != INVALID_NODE) this->free_list.push_back(n.right);
+		n.left = n.right = INVALID_NODE;
+
+		/* Recursively free the nodes being collected */
+		for (size_t i = first_free; i < this->free_list.size(); i++) {
+			node &fn = this->nodes[this->free_list[i]];
+			subtree_elements.push_back(fn.element);
+			if (fn.left != INVALID_NODE) this->free_list.push_back(fn.left);
+			if (fn.right != INVALID_NODE) this->free_list.push_back(fn.right);
+			fn.left = fn.right = INVALID_NODE;
+		}
+
+		return subtree_elements;
+	}
+
+	bool RemoveRecursive(const T &element, size_t &node_idx, int level)
+	{
+		/* Node reference */
+		auto &n = this->nodes[node_idx];
+
+		if (n.element == element) {
+			/* Remove this one */
+			this->free_list.push_back(node_idx);
+			if (n.left == INVALID_NODE && n.right == INVALID_NODE) {
+				/* Simple case, leaf, new child node for parent is "none" */
+				node_idx = INVALID_NODE;
+				return true;
+			} else {
+				/* Complex case, rebuild the sub-tree */
+				std::vector<T> subtree_elements = this->FreeSubtree(node_idx);
+				node_idx = this->BuildSubtree(subtree_elements.begin(), subtree_elements.end(), level);
+				return true;
+			}
+		} else {
+			/* Search in a sub-tree */
+			/* Dimension index of current level */
+			int dim = level % 2;
+			/* Coordinate of element splitting at this node */
+			CoordT nc = this->xyfunc(n.element, dim);
+			/* Coordinate of the element being removed */
+			CoordT ec = this->xyfunc(element, dim);
+			/* Which side to remove from */
+			size_t next = (ec < nc) ? n.left : n.right;
+			assert(next != INVALID_NODE); // node must exist somewhere and must be found before a leaf is reached
+			/* Descend */
+			if (this->RemoveRecursive(element, next, level + 1)) {
+				/* Vector may have been reallocated at this point, n and next are invalid */
+				auto &nn = this->nodes[node_idx];
+				if (ec < nc) nn.left = next; else nn.right = next;
+			}
+			return false;
+		}
+	}
+
+
+	DistT ManhattanDistance(const T &element, CoordT x, CoordT y) const
+	{
+		return abs((DistT)this->xyfunc(element, 0) - (DistT)x) + abs((DistT)this->xyfunc(element, 1) - (DistT)y);
```

Comment:
If the return value of `xyfunc` is in `CoordT` as well, one casting to `DistT` might be enough.